### ISSUE TITLE
Possibility of relaxing (removing?) internal fences

### ISSUE THREAD

**Issue by msimberg** [kokkos/kokkos#2862]
Very much related to #2860, but not triggered by it.

We've been working on a thin layer on top of Kokkos that returns HPX futures from `parallel_for/reduce/scan` and `deep_copy`. For the HPX backend this is simple. For CUDA we're attaching callbacks to streams to signal readiness of futures. Other backends will be dealt with when needed. For this to be useful we'd like to use the returned futures for necessary synchronization instead of fencing internally in Kokkos.

If one does something like this:

```
Kokkos::Cuda instance(stream);
deep_copy(instance, ...); // Host to device
parallel_X(instance, ...);
deep_copy(instance, ...); // Device to host
```

then the `deep_copy` and, depending on the arguments, `parallel_X` will fence at least after executing, if not before. Fencing (or `future.wait()`, which would have the same effect) only once at the end would be enough.

We're going to play around with removing fences for our own use anyway. I understand that this will never be the default. This will put the burden on the user (or the thin API on top of Kokkos) to make sure views live long enough etc. However, **would you accept a PR which adds an option to disable internal fences?** Has this ever come up in other contexts?

**Comment by crtrott** [comment]
parallel_x does not fence if you use the LightWeight hint, and if you use views as return arguments.
The deep_copy(instance,...) thing we need to fix.

**Comment by crtrott** [comment]
I.e.: 
```c++
Kokkos::View<double> result("Result");
//fences here
auto instance = ExecutionSpace();
auto policy = Kokkos::Experimental::require(
          Kokkos::RangePolicy<>(instance,0, N),
          Kokkos::Experimental::WorkItemProperty::HintLightWeight),
      f);
deep_copy(instance,a,b);
parallel_for(policy, functor);
parallel_reduce(policy, functor_reduce,result);
deep_copy(instance, b,a);
```
This isn't supposed to fence at all, and I think the only fence here right now (because of the reported issue) is in deep_copy(). We gonna fix that. But after that this should work.

**Comment by msimberg** [comment]
Excellent, thanks for the reply. Yes, `deep_copy` is what I was mainly getting at. Do I read this correctly then that `deep_copy` with an execution space instance will not fence at all once it's fixed, or will it just fence less (compared e.g. the case in #2860)? If I remember correctly there were some cases of `parallel_scan` that also fence. Does `HintLightWeight` disable those fences as well? Is #2860 the issue for tracking this?

**Comment by crtrott** [comment]
`deep_copy(exec,a,b)` only guarantees that the copy happens after any other work already submitted to `exec`, and before any new work submitted to `exec` after the `deep_copy` call returns. It does not guarantee that the copy has happened (or even started) by the time the call returns. Essentially it is supposed to be equivalent to enqueue a `deep_copy` operation into the execution queue of `exec`. Now it might not do that right now, because we don't test for it, but that is something we can fix. 

Regarding the scan operation it might actually but I have to test that.

**Comment by msimberg** [comment]
Thanks @crtrott for the fencing fixes for CUDA! That helps already for all use cases we have at the moment. I don't know how you prefer to track these things. From my side this could be closed now because our immediate problems are solved. We will likely want similar behaviour for HIP as well at some point. Feel free to close this if you prefer to track that elsewhere (I'll open a new issue or have a go at fixing it myself if there's nothing else by the time we get to that point ;)).

**Comment by PhilMiller** [comment]
I think we've largely fixed up the most egregious of the excessive fences, and have unit tests checking for such events around various API calls. So, closing this as done. New issues can be opened for particular cases that still fence but shouldn't.
